test/declarative_test_fwk/model.py

```diff
@@ -0,0 +1,180 @@
+    }, total=False)
+
+
+class Step:
+    """
+    Represents a declarative test step
+    """
+
+    def __init__(self, resource_directory: Path, config: StepDict, custom_resource_details: dict, replacements: dict = {}):
+        self.config = config
+        self.custom_resource_details = custom_resource_details
+        self.replacements = replacements
+
+        self.verb = None
+        self.input_data = {}
+        self.expectations = None
+
+        # (k8s.CustomResourceReference, ko) to teardown
+        self.teardown_list = []
+
+        for verb in list(Verb):
+            if verb.name not in self.config:
+                continue
+            self.verb = verb
+            self.input_data = self.config.get(verb.name)
+            if type(self.input_data) is str:
+                if self.input_data.endswith(".yaml"):
+                    # load input data from resource file
+                    resource_file_name = self.input_data
+                    resource_file_path = Path(join(resource_directory, resource_file_name))
+                    self.input_data = load_resource_file(
+                        resource_file_path.parent, resource_file_path.stem, additional_replacements=replacements)
+                else:
+                    # consider the input as resource name string
+                    # confirm that self.custom_resource_details must be provided with same name
+                    if self.custom_resource_details["metadata"]["name"] != self.input_data:
+                        raise ValueError(f"Unable to determine input data for '{self.verb}' at step: {self.id()}")
+                    # self.custom_resource_details will be mixed in into self.input_data
+                    self.input_data = {}
+            break
+
+        if len(self.input_data) == 0 and not self.custom_resource_details:
+            raise ValueError(f"Unable to determine custom resource at step: {self.id()}")
+
+        if self.custom_resource_details:
+            self.input_data = {**self.custom_resource_details, **self.input_data}
+
+        self.wait = self.config.get("wait")
+        self.expectations = self.config.get("expect")
+
+    def id(self) -> str:
+        return self.config.get("id", "")
+
+    def description(self) -> str:
+        return self.config.get("description", "")
+
+    def resource_kind(self) -> str:
+        return self.input_data.get("kind")
+
+    def __str__(self) -> str:
+        return f"Step(id='{self.id()}')"
+
+    def __repr__(self) -> str:
+        return str(self)
+
+# fields for 'customResourceReference' field in a test Scenario
+ResourceReferenceDict = TypedDict("CreateDict", {
+        "apiVersion": str,
+        "kind": str,
+        "metadata": Dict
+    }, total=False)
+
+# fields in a test scenario
+ScenarioDict = TypedDict("ScenarioDict", {
+        "id": str,
+        "description": str,
+        "marks": List[str],
+        "usecases": List[str],
+        "customResourceReference": ResourceReferenceDict,
```

Comment:
nit: I think "resource" is fine for the key here instead of the longer "customResourceReference". There's no information lost in the briefer name.

Reply:
Renamed to "resource"